### ISSUE TITLE
Feature Request: Automatic update if found new version

### ISSUE THREAD

**Issue by MetaMemoryT** [sqlitebrowser/sqlitebrowser#162]
If software can automatically check for updates, we should have a button [Update] that downloads new exe and then starts installer.

**Comment by samir-aguiar** [comment]
Something that I thought:
The program could check for updates when initializing, and...
- On Windows it would download and run the installer, that should remove the previous version and install the new files.
- On Mac perhaps it's better to integrate Sparkle into the project and let it do the job. It's widely used and probably has less bugs than the alternatives.
- On Linux I think the package manager should take care of this.

**Comment by justinclift** [comment]
Yeah, that sounds like the right approach for Windows.  For OSX, it **might** not be that straightforward.  DB Browser 4 SQLite (we really need to think of an acronym for this program :wink:) can be installed using one of the OSX package managers (eg Homebrew).  So using a solution blindly on that platform probably isn't the right approach.  [Sparkle](http://sparkle-project.org) does sound worth investigating though, for the "updating manually installed OSX binaries" use case. :smile:

Hoping your next Pull Request will be for this auto-updating stuff? :grinning:

**Comment by justinclift** [comment]
Looking at the Sparkle instructions, this seems doable:

  https://github.com/sparkle-project/Sparkle/wiki

Sounds like it's mostly a packaging approach, and won't really touch the executable code itself.  This would probably work well for our existing .dmg/.app bundle downloads, without conflicting with the Homebrew installed versions.

**Comment by samir-aguiar** [comment]
Yeah, I was thinking about giving it a try. :smile: 
But I didn't know that there was a Homebrew formula for DB Browser 4 SQLite (I'm waiting for that anagram). Homebrew doesn't officially support autoupdate, does it? If not, Mac users who wanted that would need to set up a cron job or schedule the execution of a AppleScript/Automator Task.

Anyway, I'll do some testings with Sparkle and tell what I find.

**Comment by justinclift** [comment]
If it helps, I can provide the cronjob script used for the nightly builds of D2B4S (acronym test 1 :wink:) for OSX.  Probably chuck it in a gist or something.  Useful?

**Comment by justinclift** [comment]
With Homebrew, I don't *think* it has an auto-update capability yet (unless recently added).  So yeah, people need to manually update when they're using that.

**Comment by samir-aguiar** [comment]
Yeah, it would be great. Please send it to me :grinning:

**Comment by justinclift** [comment]
Cool.  This is the script that generates the nightlies:

&nbsp; &nbsp; https://gist.github.com/justinclift/38887c1cfe480adfb11b

It's not at all elegant, and could be improved a lot (eg checking return codes for failures and similar)... but it works.

**Comment by justinclift** [comment]
Btw, I'm kinda thinking "DB4S" is the shortest potential acronym we could use...

**Comment by samir-aguiar** [comment]
Nice @justinclift, I was already thinking about asking you if you used qmake and macdeployqt to generate the bundles and .dmg files because I have another pull request to open related to the translations thing and bundle generation. Thanks!

DB4S sounds good to me :smile:

**Comment by samir-aguiar** [comment]
Just a status update: I've managed to integrate Sparkle into the app, but not without adding code. I needed to add a small Object-C++ source file to call the Sparkle methods. Sparkle also requires packages to be signed and does a bunch of checks to avoid tampered updates so the .plist files required some changes as well.
Anyway, it was just the basic integration; I still need to look at Sparkle's code to see what options they provide and run more tests.

**Comment by justinclift** [comment]
Cool.  All good progress.  Pleae keep it up. :grinning:

**Comment by justinclift** [comment]
As useful reference info, @revolter pointed out [fervor](https://github.com/pypt/fervor).  That project itself appears to have died, but it may be worth looking over its code if we need to make our own system.

**Comment by revolter** [comment]
Sparkle is great, but AFAIK is not cross platform :( `fervor` was inspired by Sparkle, and made it cross platform, so maybe we have can salvage something from there.

**Comment by justinclift** [comment]
Yep.  Our ideal solution on Windows and OSX would probably be to copy the approach of modern browsers. eg download updates automatically in the background, and install them when the application is next used.  The BSD's and Linux distro's have their own update mechanisms already, so it wouldn't need to run on those.

**Comment by chrisjlocke** [comment]
Parsing version numbers is a pain.  v3.9 is always written as 3.9 instead
of 3.09, which it is technically.  v3.10 isn't 'higher' than 3.9 (which in
maths terms is 3.90, so higher than 3.10).  Argh.  You have to break each
part of the number down rather than treating it as a 'number'.
I'm sure this is like the advice they give for logins on websites - always
take someone elses code and never roll your own! :wink:

**Comment by justinclift** [comment]
Hmmm, version number parsing for semantic versions major.minor.patch is pretty easy.  But yeah, it has to be broken into the separate parts first. :wink:

Auto-updating... yeah... we're need to have more confidence in our releases being very stable before we start doing that for everyone.  Optionally enabled (defaulting to off) code, for the brave few, is a reasonable idea to get things started with though. :smile:

**Comment by revolter** [comment]
@chrisjlocke, https://regex101.com/r/bB0hB2/1.

**Comment by justinclift** [comment]
Added v3.9.0-beta1: https://regex101.com/r/bB0hB2/2

**Comment by chrisjlocke** [comment]
v3.9.0.-1 :wink:

**Comment by justinclift** [comment]
😏

**Comment by revolter** [comment]
https://regex101.com/r/bB0hB2/3

**Comment by innermous** [comment]
@revolter it is supported. In QRegularExpression.

**Comment by revolter** [comment]
Searching for `behind` in the [`QRegularExpression`][1] documentation yields nothing :| Can you point me where is it mentioned?

If it is supported, [this regex][2] works too.

  [1]: http://doc.qt.io/qt-5/qregularexpression.html
  [2]: https://regex101.com/r/bB0hB2/5

**Comment by innermous** [comment]
@revolter I got this information [here](https://dangelog.wordpress.com/2012/04/07/qregularexpression/), it's Qt dev blog. Ctrl+F `lookbehind`, further investigations are up to you :)

**Comment by MKleusberg** [comment]
Closing this because it's really old and we have a link to the installer in the update notification already. I'm not sure either whether we should implement automatic updates at all anyway because it's just an ideal source for malware intrusion if you are not super careful. If anybody has strong feelings towards an automatic update routine, you're always welcome to re-open this issue :smile:
